engine/access/rest/converter.go

```diff
@@ -2,22 +2,219 @@ package rest
+	if err != nil {
+		return flow.Identifier{}, fmt.Errorf("invalid ID: %w", err)
+	}
+	return flowID, nil
+}
+
+func toAddress(address string) (flow.Address, error) {
+	valid, _ := regexp.MatchString(`^[0-9a-fA-F]{16}$`, address)
+	if !valid {
+		return flow.Address{}, errors.New("invalid address")
+	}
+
+	return flow.HexToAddress(address), nil
+}
+
+func toProposalKey(key *generated.ProposalKey) (flow.ProposalKey, error) {
+	address, err := toAddress(key.Address)
+	if err != nil {
+		return flow.ProposalKey{}, err
+	}
+
+	return flow.ProposalKey{
+		Address:        address,
+		KeyIndex:       uint64(key.KeyIndex),
+		SequenceNumber: uint64(key.SequenceNumber),
+	}, nil
+}
+
+func toSignature(signature string) ([]byte, error) {
+	signatureBytes := []byte(signature)
+	if len(signatureBytes) > maxSignatureLength {
+		return nil, fmt.Errorf("invalid signature")
+	}
+	return signatureBytes, nil
+}
+
+func toTransactionSignature(transactionSignature *generated.TransactionSignature) (flow.TransactionSignature, error) {
+	address, err := toAddress(transactionSignature.Address)
+	if err != nil {
+		return flow.TransactionSignature{}, err
+	}
+
+	signature, err := toSignature(transactionSignature.Signature)
+	if err != nil {
+		return flow.TransactionSignature{}, err
+	}
+
+	return flow.TransactionSignature{
+		Address:     address,
+		SignerIndex: int(transactionSignature.SignerIndex),
+		KeyIndex:    uint64(transactionSignature.KeyIndex),
+		Signature:   signature,
+	}, nil
+}
+
+func toTransactionSignatures(sigs []generated.TransactionSignature) ([]flow.TransactionSignature, error) {
+	signatures := make([]flow.TransactionSignature, len(sigs))
+	for _, sig := range sigs {
+		signature, err := toTransactionSignature(&sig)
+		if err != nil {
+			return nil, err
+		}
+
+		signatures = append(signatures, signature)
+	}
+
+	return signatures, nil
+}
+
+func toTransaction(tx *generated.TransactionsBody) (flow.TransactionBody, error) {
+
+	argLen := len(tx.Arguments)
+	if argLen > maxAllowedScriptArgumentsCnt {
+		return flow.TransactionBody{}, fmt.Errorf("too many arguments. Maximum arguments allowed: %d", maxAllowedScriptArgumentsCnt)
+	}
+
+	args := make([][]byte, argLen)
+	for _, arg := range tx.Arguments {
+		// todo validate
+		args = append(args, []byte(arg))
+	}
+
+	proposal, err := toProposalKey(tx.ProposalKey)
+	if err != nil {
+		return flow.TransactionBody{}, err
+	}
+
+	payer, err := toAddress(tx.Payer)
+	if err != nil {
+		return flow.TransactionBody{}, err
+	}
+
+	authorizerCnt := len(tx.Authorizers)
+	if authorizerCnt > maxAuthorizersCnt {
+		return flow.TransactionBody{}, fmt.Errorf("too many authorizers. Maximum authorizers allowed: %d", maxAuthorizersCnt)
+	}
+	auths := make([]flow.Address, authorizerCnt)
+	for _, auth := range tx.Authorizers {
+		a, err := toAddress(auth)
+		if err != nil {
+			return flow.TransactionBody{}, err
+		}
+
+		auths = append(auths, a)
+	}
+
+	payloadSigs, err := toTransactionSignatures(tx.PayloadSignatures)
```

Comment:
what about limits on payload/envelope signature lists? should the max be len(authorizers)+2 (payer+proposer) or something like that?

Reply:
can't the payer, propose and authorizer all do a multi sig singing with multiple keys?

Reply:
also - we have this issue to add a validation layer - https://github.com/dapperlabs/flow-go/issues/6006

Reply:
I"m not sure. I'm ok to punt on this for now and circle back later if we don't have a clear picture.

I'm taking the opportunity to hash some of this out now, but the gRPC interface has the same lack of limits so we'll eventually need to address it globally

Reply:
tbh - for many of these, I don't know what the right limits are. Will a separate issue to tackle these limits overall (grpc, web proxy and REST) help?

Reply:
https://github.com/dapperlabs/flow-go/issues/6028

Reply:
> tbh - for many of these, I don't know what the right limits are. Will a separate issue to tackle these limits overall (grpc, web proxy and REST) help?

yea, that's a good idea. we should at least get the conversation started about how to manage it